File: backend/internal/telegram/linkstore.go
package telegram

import (
	"crypto/rand"
	"encoding/base64"
	"sync"
	"time"

	"github.com/google/uuid"
)

const linkTokenTTL = 5 * time.Minute

type linkEntry struct {
	UserID    uuid.UUID
	ExpiresAt time.Time
}

// LinkStore holds pending deep-link tokens for Telegram account linking.
// Tokens expire after 5 minutes. Thread-safe.
type LinkStore struct {
	mu      sync.Mutex
	entries map[string]linkEntry
}

// NewLinkStore creates a new in-memory link token store.
func NewLinkStore() *LinkStore {
	return &LinkStore{
		entries: make(map[string]linkEntry),
	}
}

// GenerateToken creates a new opaque token that maps to the given userID.
// Returns a URL-safe base64 string (32 bytes of randomness).
func (s *LinkStore) GenerateToken(userID uuid.UUID) (string, error) {
	b := make([]byte, 32)
	if _, err := rand.Read(b); err != nil {
		return "", err
	}
	token := base64.RawURLEncoding.EncodeToString(b)

	s.mu.Lock()
	s.entries[token] = linkEntry{
		UserID:    userID,
		ExpiresAt: time.Now().Add(linkTokenTTL),
	}
	s.mu.Unlock()

	return token, nil
}

// ConsumeToken looks up the token, returns the associated userID if valid
// and not expired, and deletes the entry (one-shot). Returns false if the
// token is unknown or expired.
func (s *LinkStore) ConsumeToken(token string) (uuid.UUID, bool) {
	s.mu.Lock()
	defer s.mu.Unlock()

	entry, exists := s.entries[token]
	if !exists {
		return uuid.Nil, false
	}

	delete(s.entries, token)

	if time.Now().After(entry.ExpiresAt) {
		return uuid.Nil, false
	}

	return entry.UserID, true
}

File: backend/src/telegram/linkstore.rs
use std::collections::HashMap;
use std::time::{Duration, Instant};

use base64::Engine;
use parking_lot::Mutex;
use rand::RngCore;
use uuid::Uuid;

const LINK_TTL: Duration = Duration::from_secs(5 * 60);

struct Entry {
    user_id: Uuid,
    expires_at: Instant,
}

#[derive(Default)]
pub struct LinkStore {
    entries: Mutex<HashMap<String, Entry>>,
}

impl LinkStore {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn generate_token(&self, user_id: Uuid) -> String {
        let mut bytes = [0u8; 32];
        rand::thread_rng().fill_bytes(&mut bytes);
        let token = base64::engine::general_purpose::URL_SAFE_NO_PAD.encode(bytes);
        self.entries.lock().insert(
            token.clone(),
            Entry {
                user_id,
                expires_at: Instant::now() + LINK_TTL,
            },
        );
        token
    }

    /// Consume the token. Returns `Some(user_id)` if valid and unexpired;
    /// otherwise returns `None`. The entry is always removed on lookup.
    pub fn consume_token(&self, token: &str) -> Option<Uuid> {
        let mut guard = self.entries.lock();
        let entry = guard.remove(token)?;
        if Instant::now() > entry.expires_at {
            None
        } else {
            Some(entry.user_id)
        }
    }
}
